[21.09] Limit lifetime of repository cache

mvdbeek opened the pull request:
Another attempt at fixing https://github.com/galaxyproject/galaxy/issues/12988

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
Limit lifetime of repository cache

jdavcs commented on lib/galaxy/tools/cache.py:
```diff
@@ -288,20 +287,20 @@ class ToolShedRepositoryCache:
         self.repos_by_tuple = defaultdict(list)
-        self.rebuild()
+        self._rebuild()
+        self.session.close()
```
So we have a `defer` clause in `_rebuild`..  Since `session.close()` expunges the session,  you should get a DetachedInstanceError when trying to access the `ToolShedRepository.metadata` property. If we call `_rebuild` before accessing the cache, then it will repopulate the session, so that's fine. So this would work only if:
a) we do not access the `metadata` property, or
b) we call `_rebuild` before accessing it.

I'm not sure about the context in which we access the cache, but would it make sense to *not* call `self._rebuild()` from the constructor, and unconditionally rebuild on first access? (that could be done with a property that would lazy-load `self.repositories` maybe?).

mvdbeek replied:
It'll never be accessed, the cache is disposed before the app startup is finished.

mvdbeek commented on lib/galaxy/tools/cache.py:
```diff
@@ -288,20 +287,20 @@ class ToolShedRepositoryCache:
+    def _rebuild(self):
         self.repositories = self.session.query(ToolShedRepository).options(
             defer(ToolShedRepository.metadata), joinedload('tool_dependencies')
         ).all()
```
```suggestion
        self.repositories = self.session.query(ToolShedRepository).options(
            defer(ToolShedRepository.metadata)
        ).all()
```

mvdbeek committed:
Don't need to use a joinedload anymore

The cache is only used on startup, not during dependency resolution.
Also rename _rebuild, since we only use it once now.

nsoranzo commented on lib/galaxy/tool_shed/util/repository_util.py:
```diff
@@ -208,15 +208,17 @@ def get_installed_repository(app, tool_shed=None, name=None, owner=None, changes
     """
```
Should we remove `refresh=False` from the function arguments a couple of lines above?

mvdbeek committed:
Drop refresh argument of get_installed_repository
